frontend/src/poc-code/testK8s/ws/WSFactory.ts

```diff
@@ -0,0 +1,338 @@
+  }
+
+  private connect() {
+    this.state = 'init';
+    this.messageBuffer = [];
+
+    const url = createURL(applyConfigHost(this.options.host), this.options.path);
+    const subProtocols = applyConfigSubProtocols(
+      this.options.host ? this.options.subProtocols : undefined,
+    );
+    try {
+      this.ws = new WebSocket(url, subProtocols);
+    } catch (e) {
+      // eslint-disable-next-line no-console
+      console.error('Error creating websocket:', e);
+      this.reconnect();
+      return;
+    }
+
+    this.ws.onopen = () => {
+      // eslint-disable-next-line no-console
+      console.log(`websocket open: ${this.id}`);
+      this.state = 'open';
+      this.triggerEvent('open', undefined);
+      if (this.connectionAttempt) {
+        clearTimeout(this.connectionAttempt);
+        this.connectionAttempt = null;
+      }
+    };
+    this.ws.onclose = (evt) => {
+      // eslint-disable-next-line no-console
+      console.log(`websocket closed: ${this.id}`, evt);
+      this.state = 'closed';
+      this.triggerEvent('close', evt);
+      this.reconnect();
+    };
+    this.ws.onerror = (evt) => {
+      // eslint-disable-next-line no-console
+      console.log(`websocket error: ${this.id}`);
+      this.state = 'error';
+      this.triggerEvent('error', evt);
+    };
+    this.ws.onmessage = (evt) => {
+      const msg = this.options?.jsonParse ? JSON.parse(evt.data) : evt.data;
+      // In some browsers, onmessage can fire after onclose/error. Don't update state to be incorrect.
+      if (this.state !== 'destroyed' && this.state !== 'closed') {
+        this.state = 'open';
+      }
+      this.triggerEvent('message', msg);
+    };
+  }
+
+  /**
+   * Invoke all registered handler callbacks for a given data.
+   */
+  private invokeHandlers(type: EventHandlerTypes, data?: unknown) {
+    const handlers = this.handlers[type];
+    if (!handlers) {
+      return;
+    }
+    handlers.forEach((h) => {
+      try {
+        // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+        // @ts-ignore
+        h(data); // typescript is having an issue with passing the data, muting for now
+      } catch (e) {
+        // eslint-disable-next-line no-console
+        console.error('WS handling failed:', e);
+      }
+    });
+  }
+
+  /**
+   * Triggers event to be buffered or invoked depending on config.
+   */
+  private triggerEvent(type: EventHandlerTypes, data?: unknown) {
+    if (this.state === 'destroyed') {
+      return;
+    }
+
+    const isMessageEvent = (t: string, e: unknown): e is MessageDataType => {
+      return t === 'message' && !!e;
+    };
+
+    // Only buffer "message" events, so "error" and "close" etc can pass thru.
+    if (this.bufferMax && isMessageEvent(type, data)) {
+      this.messageBuffer.push(data);
+
+      if (this.messageBuffer.length > this.bufferMax) {
+        this.messageBuffer.shift();
+      }
+
+      return;
+    }
+
+    this.invokeHandlers(type, data);
```

Comment:
I am not a big fan of `empty` returns, wouldn't it be better to use `else` in here?
```suggestion
    } else {
       this.invokeHandlers(type, data);
    }
```

Reply:
Okay, I think this is a matter of preference. I am fine with empty returns, it's just always "Did we forgot to return something in here"? And I have to read the whole block of code again to understand if something was not supposed to be returned.

Reply:
That's a JavaScript concern 😛 

This is TypeScript. We can type our returns (and I have in my other PR). You cannot violate the return statement without a TS error.